test_runner/src/main/kotlin/ftl/reports/api/PrepareForJUnitResult.kt

```diff
@@ -0,0 +1,45 @@
+package ftl.reports.api
+
+import com.google.api.services.toolresults.model.Step
+import com.google.api.services.toolresults.model.TestCase
+import ftl.reports.api.data.TestExecutionData
+
+// List of TestExecutionData can contains also secondary steps from flaky tests reruns.
+// We need only primary steps, but we also prefer to display failed tests over successful,
+// so we overrides successful test cases with failed from secondary steps
+internal fun List<TestExecutionData>.prepareForJUnitResult(): List<TestExecutionData> = this
+    .reduceToPrimarySteps()
+    .reduceTestCases()
+
+private fun List<TestExecutionData>.reduceToPrimarySteps(): List<TestExecutionData> = groupBy { data ->
+    data.step.primaryStepId
+}.mapNotNull { (_, list: List<TestExecutionData>) ->
+    list.sortedBy { data ->
+        data.step.multistepNumber
+    }.reduce { primary, next ->
+        primary.copy(testCases = primary.testCases + next.testCases)
+    }
+}
+
+private fun List<TestExecutionData>.reduceTestCases() = map(TestExecutionData::reduceTestCases)
+
+private fun TestExecutionData.reduceTestCases() = copy(
+    testCases = testCases.groupBy(TestCase::getTestCaseId).map { (_, testCases) ->
+        testCases.sortedBy { testCase: TestCase ->
+            testCase.startTime.asUnixTimestamp()
+        }.run {
+            firstOrNull { it.stackTraces != null } ?: first()
```

Comment:
It can be just me (and probably is) but this lines looks bit strange for me (at least for the first glance). 
```
.run {
    firstOrNull { it.stackTraces != null } ?: first()
}
```
I am wondering if it is doable and we could add private extension function like `.firstWithStackTrace` or sth similar. 
Just thinking

Reply:
As you wish, code should be readable for everyone who contribute

Reply:
I guess `firstWithStackTrace` can be little bit misleading. Should be `firstWithStackTraceOrFirst`, but is to long, and I avoid names that contain condition. So I hope this construction will solve the problem

```
.run {
    if (!isFlaky()) first()
    else first { it.stackTraces != null }.apply { flaky = true }
}
```